### DISCUSSION TITLE
Unable to run without installing

### DISCUSSION THREAD

**Discussion by jsbien** [FriedrichFroebel/didjvu#22] [Q&A] [answered]
From README.rst:
> Alternatively, you can use didjvu without installing it, straight out of an unpacked source tarball or a VCS checkout.

For the repository cloned today I get (the prompt is misleading, this is actually Debian bookworm, i.e. stable):

`jsbien@JSBbuster:~/work/didjvu/Froebel_new/didjvu/didjvu$ python3 didjvu.py`
`ImportError: cannot import name 'cli' from partially initialized module 'didjvu' (most likely due to a circular import) (/home/jsbien/work/didjvu/Froebel_new/didjvu/didjvu/didjvu.py)`
Am I doing something wrong?

**Comment by FriedrichFroebel** [comment, marked as answer]
You should not run from the `didjvu.py` file, but from the actual module when using it without installation:

1. `git clone git@github.com:FriedrichFroebel/didjvu.git`
2. `cd didjvu`
3. Install dependencies (at least `Pillow` and `gamera`).
4. `python -m didjvu --help` (assuming that you use a virtual environment with Python 3 or your default interpreter is Python 3, otherwise replace `python` with `python3`).

**Reply by jsbien** [reply]
Looks like my gamera installation somehow vanished :-(
An attempt to reinstall it with pip, pip3 and pipx fail. BTW, I get the same errors when trying to install directly didjvu with pip, pip3 and pipx.

**Reply by FriedrichFroebel** [reply]
For now, Gamera has to be installed manually, due to the package on PyPI being outdated and only providing a Windows binary (see https://github.com/hsnr-gamera/gamera-4/issues/65 where we try to improve this). For installing it, just follow the manual steps:

1. `git clone git@github.com:hsnr-gamera/gamera-4.git`
2. `cd gamera-4`
3. `python -m pip install . --verbose`

**Reply by jsbien** [reply]
Thanks for your help!

> git clone git@github.com:hsnr-gamera/gamera-4.git

This doesn't work for me, but this works:
`git clone https://github.com/hsnr-gamera/gamera-4.git`
Unfortunately
`python3 -m pip install . --verbose`
gives
`error: externally-managed-environment`
...
` You can override this, at the risk of breaking your Python installation or OS, by passing --break-system-packages.`

**Reply by FriedrichFroebel** [reply]
My clone command only works if SSH is configured, otherwise you need the HTTPS version, yes.

Usually, you want to do everything in a virtual environment:

1. Create it: `python3 -m virtualenv venv`
2. Enter it: `source venv/bin/activate`
3. Use as the global environment, for example installing packages etc.
4. To exit it: `deactivate`

You can use the global environment as well, but this is not recommended (any more), if you use the `--user` flag for *pip* or run each installation with `sudo`.

**Comment by jsbien** [comment]
Thank again for your help. Looks like my problems are solved.

BTW, instead

> python3 -m virtualenv venv

there should be e.g.
`python3 -m venv venv`

**Reply by FriedrichFroebel** [reply]
This depends on your preferences. `venv` is the stdlib module, `virtualenv` an alternative with more features: https://stackoverflow.com/questions/41573587/what-is-the-difference-between-venv-pyvenv-pyenv-virtualenv-virtualenvwrappe

**Reply by jsbien** [reply]
I will look into it. For `python3 -m virtualenv venv` I've got `/usr/bin/python3: No module named virtualenv`...

**Reply by FriedrichFroebel** [reply]
Depending on your setup, you will have to install the corresponding OS or global Python package: https://packages.debian.org/bookworm/python3-virtualenv https://pypi.org/project/virtualenv/
